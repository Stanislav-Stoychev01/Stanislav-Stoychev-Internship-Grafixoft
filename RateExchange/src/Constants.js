export const currencyList = ['USD', 'EUR', 'AUD', 'CAD', 'CHF', 'NZD', 'BGN'];

export const defaultCurrency = 'USD';


export const columnCount = 3;
export const firstColumn = 'First Column';
export const secondColumn = 'Second Column';
export const thirdColumn = 'Third Column';

export const firstColumnIndex = 0;
export const secondColumnIndex = 1;
export const thirdColumnIndex = 2;

export const key = 0;
export const value = 1;

export const dateString = 'date';
export const latest = 'latest';

export const inputLabel = "Input Currency Name:";
export const exchangeRateLabel = "Exchange Rates from";
export const longestSequenceMessage = "The longest sequence has length of ";

export const tableColumnNames = ['Conversion Rates below 1', 
'Conversion Rates between 1 and 1,5', 'Conversion Rates above 1,5'];

export const LoadingData = 'The data is currenly loading';