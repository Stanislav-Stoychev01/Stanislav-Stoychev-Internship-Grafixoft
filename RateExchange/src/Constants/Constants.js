export const CurrencyList = ['USD', 'EUR', 'AUD', 'CAD', 'CHF', 'NZD', 'BGN']
export const DefaultCurrency = 'USD'
