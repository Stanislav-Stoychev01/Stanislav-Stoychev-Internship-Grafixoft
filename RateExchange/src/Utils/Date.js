export const getCurrentDay = () => {

    const date = new Date();
    const getDay = String(date.getDate()).padStart(2, '0');
    const getMonth = String(date.getMonth() + 1).padStart(2, '0');
    const getYear = date.getFullYear();
    const currentDate = getYear + '-' + getMonth + '-' + getDay;

    return currentDate;
}

export default getCurrentDay