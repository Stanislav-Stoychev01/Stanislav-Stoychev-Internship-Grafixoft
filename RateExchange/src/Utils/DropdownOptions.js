export const dropdownOptionsData = (array) => {

    const dropdownAraray = [];

    array.forEach((item, index) => {
        const dropdownArarayItem = {
            value: array[index],
            label: array[index],
        };

        dropdownAraray.push(dropdownArarayItem);
    });

    return dropdownAraray;
}

export default dropdownOptionsData;