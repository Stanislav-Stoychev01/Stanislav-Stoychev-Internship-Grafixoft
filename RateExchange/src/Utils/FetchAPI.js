
export const FetchCurrency = (currency, date) => {

    return (
        fetch(`https://cdn.jsdelivr.net/gh/fawazahmed0/currency-api@1/${date}/currencies/${currency}.json`)
        .then((res) => res.json())
    );
}

export default FetchCurrency