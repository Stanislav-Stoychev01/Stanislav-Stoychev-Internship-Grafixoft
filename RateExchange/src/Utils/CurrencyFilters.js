import * as Constants from '../Constants';

export const getSpecifiedCurrencies = (fetchedCurrencyArray) => {

    return fetchedCurrencyArray.filter((currency) => isCurrencyValid(currency));
}

const isCurrencyValid = (currency) => {

    const [ currencyName, currencyValue ] = currency
    const { currencyList } = Constants;
    let result = false;
   
    currencyList.forEach(element => {
        const lowerCaseCurrency = element.toLowerCase();
        if (lowerCaseCurrency === currencyName
            && currencyValue !== 1) {
            result = true;
        }
    });

    return result;
}

export const sortCurrencyArray = (currencyArray) => { 

    const currencyValueIndex = Constants.value;

    return currencyArray.sort((currency, nextCurrency) => 
    currency[currencyValueIndex] - nextCurrency[currencyValueIndex]);
} 

export const sortIntoColumns = (currencies) => {

    const tableColumns = [[], [], []];

    currencies.forEach((element) => {
        const columnIndex = determineColumn(element);
        tableColumns[columnIndex].push(element);
    })
    return tableColumns;
}

export const determineColumn = (keyValueItem) => {

    const value = keyValueItem[Constants.value];
    
    let result;

    switch(true) {
        
        case (Number(value) < Number(1)): {
            result = 0;
            break;
        }
        case (Number(value) > Number(1) && Number(value) < Number(1.5)): {
            result = 1
            break;
        }
        case (Number(value) > Number(1.5)): {
            result = 2;
            break;
        }
    }

    return result;
}

const findColumnsLengths = (tableColumnsArray) => {

    return tableColumnsArray.map((column) => column.length);
}

export const convertToRows = (columnsArray) => {

    const result = [];
    const columnLengthsArray = findColumnsLengths(columnsArray);
    const rowsSize = Math.max(...columnLengthsArray);
    const colsSize = columnLengthsArray.length;

    for(let row = 0; row < rowsSize; row++)
    {
        const rowArray = Array.apply(undefined, Array(colsSize));

        for(let col = 0; col < colsSize; col++)
        {
            const currentItem = columnsArray[col][row];

            if (currentItem) {
                const determinePosition = determineColumn(currentItem);
                rowArray[determinePosition] = currentItem;
            }          
        }

        result.push(rowArray);
    }

    return result;
}

export const sortKeyValues = (response) => {

    const dataCurrency = Object.entries(response);
    const currenctyList = getSpecifiedCurrencies(dataCurrency);
    const sortedCurrenctyList = sortCurrencyArray(currenctyList);

    return sortedCurrenctyList;
}

export const applyFilters = (response) => {

    const sortedList = sortKeyValues(response);
    const sortIntoCols = sortIntoColumns(sortedList);  

    return sortIntoCols;
}
