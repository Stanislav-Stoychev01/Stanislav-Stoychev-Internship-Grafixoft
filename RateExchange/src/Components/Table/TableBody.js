
import * as CurrencyFilters from '../../Utils/CurrencyFilters';


function TableBody(props) {

    const TableData = () => {
        return (
            CurrencyFilters.convertToRows(props.data).map((array, arrayIndex) => {
                return (
                    <tr key={arrayIndex} className="table-row">
                        {
                            array.map((item, itemIndex) => {
                                const undefinedItem = [undefined, undefined];
                                const [firstItem, secondItem] = item ?? undefinedItem;
                                return (
                                    <td key={itemIndex} className="table-col">
                                        {firstItem} {secondItem}
                                    </td>
                                )
                            })
                        }
                    </tr>
                )
            })
        )
    }

    const TableColumnsLenghts = () => {
        return (
            <tr className="table-row">
                {
                    props.data.map((value, index) => {
                        return (
                            <td key={index} className="table-col">
                                { props.label } {value.length}
                            </td>
                        )
                    })
                }
            </tr>
        )
    }

    return (
        props.data !== undefined
        && <tbody id="table-body">
            <TableData />
            <TableColumnsLenghts />
        </tbody>
    )

}

export default TableBody