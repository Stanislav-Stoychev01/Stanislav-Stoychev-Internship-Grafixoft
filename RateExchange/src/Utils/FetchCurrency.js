import { fetchAPI } from "../API/FetchAPI";
import * as LocalStorage from "../Utils/LocalStorage";
import * as Constants from "../Constants";
import { getCurrentDay } from "../Utils/Date";

export async function fetchCurrency(currency, date) {

    const currentDate = getCurrentDay();
    const lowerCaseCurrency = currency.toLowerCase();
    const currentDateKey = currentDate + ' ' + lowerCaseCurrency;
    const currentDateRecord = localStorage.getItem(currentDateKey);
    let response;

    if (date === undefined) {
        date = Constants.latest;
    }

    const currentRecordKey = date + ' ' + lowerCaseCurrency;  
    const currentRecord = localStorage.getItem(currentRecordKey);
    

    if (currentRecord === null) {

        if (currentDateRecord !== null) {
            response = JSON.parse(currentDateRecord);
        }

        else {
            response = await fetchAPI(currency, date);
            LocalStorage.removeCurrencyData(currency);
            const responseToJson = JSON.stringify(response);
            localStorage.setItem(currentRecordKey, responseToJson);
        }
    }
    else {
        response = JSON.parse(currentRecord);
    }

    return response;

}

export default fetchCurrency