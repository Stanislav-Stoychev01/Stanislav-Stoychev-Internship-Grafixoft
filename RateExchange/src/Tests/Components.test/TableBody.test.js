import TableBody from "../../Components/Table/TableBody";
import * as Constants from "../../Constants";
import * as CurrencyFilters from "../../Utils/CurrencyFilters"
import Enzyme from 'enzyme';
import toJson from "enzyme-to-json";
import { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

Enzyme.configure({ adapter: new Adapter() });

it("renders table data when all props are provided correctly", () => {
    
    const response = {  
        'cad': '1.759853',
        'aud': '0.014107',
        'bgn': '2.618904',
        'eur': '3.673103',
        'afn': '85.586079'
    };

    let currencyName = 'usd'

    const data = CurrencyFilters.applyFilters(response);

    const { tableColumnsLenghtsLabel } = Constants;

    const props = {
        data: data,
        countField: tableColumnsLenghtsLabel,
    };

    const wrapper = shallow(<TableBody {...props}/>);

    const jsonWrapper = toJson(wrapper);

    expect(jsonWrapper).toMatchSnapshot();
})