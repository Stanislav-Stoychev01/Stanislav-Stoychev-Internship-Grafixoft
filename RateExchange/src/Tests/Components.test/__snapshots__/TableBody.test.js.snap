// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders table data when all props are provided correctly 1`] = `
<tbody
  id="table-body"
>
  <TableData />
  <TableColumnsLenghts />
</tbody>
`;
